xonsh/ptk_shell/formatter.py

```diff
@@ -0,0 +1,45 @@
+import builtins
+import functools
+from concurrent.futures import ThreadPoolExecutor
+
+from xonsh.prompt.base import PromptFormatter, DEFAULT_PROMPT
+
+
+class ThreadedPromptFormatter(PromptFormatter):
+    def __call__(self, template=DEFAULT_PROMPT, fields=None, threaded=False):
+        """Formats a xonsh prompt template string."""
+        self.template = template  # current template
+        self.threaded = threaded
+        self.futures = {}
+        if threaded and (not hasattr(self, "thread_pool")):
```

Comment:
Wouldn't it be cheaper to set this on `__init__` rather than checking with `hasattr` for each call?

Reply:
the idea is to support both threaded and synchronous calls. So if we put this on init, it might be occupying unnecessary memory for sync prompts. I will change the class name of the formatter to be like PTKFormatter or something.